File: server/contracts/discovery.py
"""
Discovery Contract - Unified schema for Facts, Index, and Files discovery.

This module defines the canonical data models for unified discovery across
all three domains (Facts, Index, Files) with consistent deep-linking support.
"""
from typing import List, Optional, Dict, Literal, Any
from pydantic import BaseModel, Field
from datetime import datetime


class DiscoveryQuery(BaseModel):
    """Query parameters for unified discovery search."""
    query: str = Field(..., description="Search query string")
    scope: List[Literal["facts", "index", "files"]] = Field(
        default=["facts", "index", "files"],
        description="Which domains to search (default: all)"
    )
    limit: int = Field(default=20, ge=1, le=100, description="Maximum hits per domain")
    offset: int = Field(default=0, ge=0, description="Offset for pagination")
    chat_id: Optional[str] = Field(None, description="Optional chat ID filter")
    project_id: Optional[str] = Field(None, description="Project ID (required for Facts/Index)")
    filters: Optional[Dict[str, Any]] = Field(
        None,
        description="Future: additional domain-specific filters"
    )


class DiscoverySource(BaseModel):
    """
    Source metadata for deep linking and provenance.
    
    Each hit can have multiple sources (e.g., a fact from a chat message,
    or a file chunk from a specific file).
    """
    kind: Literal["chat_message", "file", "fact"] = Field(
        ...,
        description="Type of source: chat_message, file, or fact"
    )
    
    # Chat message deep linking
    source_message_uuid: Optional[str] = Field(
        None,
        description="Stable UUID of the chat message (for deep linking to chat messages)"
    )
    source_chat_id: Optional[str] = Field(
        None,
        description="Chat ID containing the message (for navigation context)"
    )
    
    # File deep linking
    source_file_id: Optional[str] = Field(
        None,
        description="File ID (for deep linking to files)"
    )
    source_file_path: Optional[str] = Field(
        None,
        description="File path relative to source root (for deep linking to files)"
    )
    
    # Fact deep linking
    source_fact_id: Optional[str] = Field(
        None,
        description="Fact ID (for fact-specific deep linking)"
    )
    
    # Additional metadata
    snippet: Optional[str] = Field(
        None,
        description="Snippet of source content for preview"
    )
    created_at: Optional[str] = Field(
        None,
        description="ISO format timestamp when source was created"
    )
    meta: Dict[str, Any] = Field(
        default_factory=dict,
        description="Additional domain-specific metadata"
    )


class DiscoveryHit(BaseModel):
    """
    A single discovery result from any domain (Facts, Index, or Files).
    
    All hits share the same structure for unified rendering and ranking.
    """
    id: str = Field(
        ...,
        description="Stable per-hit ID: facts:<fact_id>, index:<message_uuid>:<chunk_id>, files:<file_id>:<chunk_id>"
    )
    domain: Literal["facts", "index", "files"] = Field(
        ...,
        description="Which domain this hit came from"
    )
    type: str = Field(
        ...,
        description="Hit type: 'fact', 'chat_chunk', 'file_chunk', 'file_metadata'"
    )
    title: Optional[str] = Field(
        None,
        description="Display title (e.g., fact key, file name, message preview)"
    )
    text: str = Field(
        ...,
        description="Primary display text (fact value, chunk content, file metadata)"
    )
    score: Optional[float] = Field(
        None,
        description="Relevance score (0.0 to 1.0, higher is better)"
    )
    rank: Optional[int] = Field(
        None,
        description="Rank within domain (for ranked lists like favorite colors)"
    )
    sources: List[DiscoverySource] = Field(
        default_factory=list,
        description="Source metadata for deep linking (typically one source per hit)"
    )
    meta: Dict[str, Any] = Field(
        default_factory=dict,
        description="Additional domain-specific metadata"
    )


class DiscoveryResponse(BaseModel):
    """
    Unified response from discovery search across all domains.
    
    Includes hits, counts, timings, and degraded status for observability.
    """
    query: str = Field(..., description="Original query string")
    hits: List[DiscoveryHit] = Field(
        default_factory=list,
        description="Merged and ranked hits from all domains"
    )
    counts: Dict[str, int] = Field(
        default_factory=dict,
        description="Per-domain hit counts: {'facts': 5, 'index': 10, 'files': 3}"
    )
    timings_ms: Dict[str, float] = Field(
        default_factory=dict,
        description="Per-domain query timings in milliseconds: {'facts': 45.2, 'index': 123.5, 'files': 67.8}"
    )
    degraded: Dict[str, str] = Field(
        default_factory=dict,
        description="Degraded status per domain: {'index': 'timeout', 'files_content': 'unavailable'}"
    )

File: web/src/api/discovery.ts
/**
 * Discovery API client for unified search across Facts, Index, and Files.
 * 
 * This module provides TypeScript types and functions matching the backend
 * Discovery Contract (server/contracts/discovery.py).
 */

export type DiscoveryDomain = "facts" | "index" | "files";
export type DiscoverySourceKind = "chat_message" | "file" | "fact";

export interface DiscoveryQuery {
  query: string;
  scope?: DiscoveryDomain[];
  limit?: number;
  offset?: number;
  chat_id?: string;
  project_id?: string;
  filters?: Record<string, any>;
}

export interface DiscoverySource {
  kind: DiscoverySourceKind;
  source_message_uuid?: string;
  source_chat_id?: string;
  source_file_id?: string;
  source_file_path?: string;
  source_fact_id?: string;
  snippet?: string;
  created_at?: string;
  meta?: Record<string, any>;
}

export interface DiscoveryHit {
  id: string;
  domain: DiscoveryDomain;
  type: string;
  title?: string;
  text: string;
  score?: number;
  rank?: number;
  sources: DiscoverySource[];
  meta?: Record<string, any>;
}

export interface DiscoveryResponse {
  query: string;
  hits: DiscoveryHit[];
  counts: Record<string, number>;
  timings_ms: Record<string, number>;
  degraded: Record<string, string>;
}

/**
 * Search across Facts, Index, and Files using the unified Discovery endpoint.
 * 
 * @param payload - Discovery query parameters
 * @returns Promise resolving to DiscoveryResponse
 */
export async function searchDiscovery(payload: DiscoveryQuery): Promise<DiscoveryResponse> {
  const response = await fetch("http://localhost:8000/discovery/search", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      query: payload.query,
      scope: payload.scope || ["facts", "index", "files"],
      limit: payload.limit || 20,
      offset: payload.offset || 0,
      chat_id: payload.chat_id,
      project_id: payload.project_id,
      filters: payload.filters,
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    let errorMessage = "Failed to search";
    try {
      const errorData = JSON.parse(errorText);
      errorMessage = errorData.detail || errorData.message || errorMessage;
    } catch {
      errorMessage = errorText || errorMessage;
    }
    throw new Error(errorMessage);
  }

  return response.json();
}

